Add liking and unliking tag functionality

JunXingLiu opened the pull request.

JunXingLiu committed:
Add liking and unliking tag

chrisK00 commented on Server/JokesOnYou.Web.Api/Controllers/LikedTagsController.cs:
```diff
@@ -11,5 +19,26 @@ public LikedTagsController(ILikedTagsService likedTagsService)
             _likedTagsService = likedTagsService;
         }
+
+        [Authorize(Roles = "Registered,Admin")]
```
Can be placed on top of the controller so we dont need to repeat it

JunXingLiu replied:
done

chrisK00 commented on Server/JokesOnYou.Web.Api/Controllers/LikedTagsController.cs:
```diff
@@ -11,5 +19,26 @@ public LikedTagsController(ILikedTagsService likedTagsService)
         }
+
+        [Authorize(Roles = "Registered,Admin")]
+        [HttpPut]
```
I would prefer to use HttpPost here, since we are creating a resource/making an action rather than fully updating an existing one. Same comments apply to the other method

JunXingLiu replied:
done

chrisK00 commented on Server/JokesOnYou.Web.Api/Models/Request/LikedTagsCreateDto.cs:
```diff
@@ -1,6 +1,9 @@
+
+namespace JokesOnYou.Web.Api.Models.Request
 {
     public class LikedTagsCreateDto
```
I think we dont need this class tho considering we only need the id which should be passed in the url something like this
api/tags/like/1

JunXingLiu replied:
done

chrisK00 commented on Server/JokesOnYou.Web.Api/Models/Tag.cs:
```diff
@@ -15,5 +15,6 @@ public class Tag
         public ICollection<int> Jokes { get; set; } = new List<int>();
         public string OwnerId { get; set; }
         public int Likes { get; set; }
+        public ICollection<User> Users { get; set; } = new List<User>();
```
Can we have a reference to the Users id by creating a join table instead of the actual entity?

JunXingLiu replied:
Sorry, I am pretty new to this, how can I do that?
https://www.entityframeworktutorial.net/code-first/configure-one-to-many-relationship-in-code-first.aspx
I used this site as a reference but I can't find how you could do it without creating the entity?

chrisK00 replied:
Correct me if im wrong I was looking for an example where we have already done this, and there seems to be a entity for this feature

https://github.com/Code2Gether-Discord/JokesOnYou/blob/dev/Server/JokesOnYou.Web.Api/Models/LikedTags.cs

Anyhow the general approach is just like how you would do a many to many in sql. You have a User table, You have a Tag table, then you have UserLikedTag table which holds a reference to the user_id and the tag_id.

HellHunterMax replied:
@JunXingLiu in Nico's PR you can see how he did it:
https://github.com/Code2Gether-Discord/JokesOnYou/pull/190/files

I see now that the entity we created will not work because it uses a list. change the list to a int, and save every tag/joke/user combo separately.

chrisK00 commented on Server/JokesOnYou.Web.Api/Services/LikedTagsService.cs:
```diff
@@ -1,15 +1,77 @@
+        public LikedTagsService(ILikedTagsRepository likedTagsRepo, ITagRepository tagRepository, IUserRepository userRepository, IUnitOfWork unitOfWork, IMapper mapper)
         {
             _likedTagsRepo = likedTagsRepo;
+            _tagRepo = tagRepository;
```
No cheating! :P _tagRepository

JunXingLiu replied:
I changed to _tagsService

JunXingLiu committed:
Merge branch 'dev' into feature/#90-user-liked-tags

JunXingLiu committed:
LikedTag functionality

-Added an intermediate table for Tag and User.
-LikedTags are added as each record.
-Delete last migration and added a new one.

HellHunterMax commented on Server/JokesOnYou.Web.Api/Services/LikedTagsService.cs:
```diff
@@ -14,60 +14,71 @@ namespace JokesOnYou.Web.Api.Services
+            if (likedTags == null)
             {
-                throw new KeyNotFoundException("Tag wasn't liked user");
+                throw new AppException("Tag wasn't liked by user");
```
Maybe add some more info in this exception, like which tag and which user?

JunXingLiu replied:
done

HellHunterMax commented on Server/JokesOnYou.Web.Api/Services/LikedTagsService.cs:
```diff
@@ -14,60 +14,71 @@ namespace JokesOnYou.Web.Api.Services
             }
 
-            if (!tagToUnlike.Users.Contains(userUnlikingTag))
+            LikedTags likedTags = await _likedTagsRepo.GetLikedTag(userId, tagID);
```
LikedTags Should be changed to TagLike or LikedTag or LikeForTag but my vote goes do the Last one.
1. its a Tag Like, (a like for a tag). Not a Liked tag (a tag that has been liked). the object is the like not the tag.
2. its singular not muliple likedTags anymore.
3. LikeForTag makes it clear that its a like for a tag.

I hope this is clear.

JunXingLiu replied:
Done, I used likeForTag.

HellHunterMax commented on Server/JokesOnYou.Web.Api/Services/LikedTagsService.cs:
```diff
@@ -14,60 +14,71 @@ namespace JokesOnYou.Web.Api.Services
         }
 
-        public async Task<LikedTagsReplyDto> LikeTagAsync(LikedTagsCreateDto likedTagCreateDto, string userId)
+        public async Task<LikedTagsReplyDto> LikeTagAsync(int tagID, string userId)
```
tagId not tagID. just a small tweak.

JunXingLiu replied:
done

HellHunterMax reviewed (changes requested):
So because Tag has not been implemented yet this is not going to work as intended.
With this you will only be able to like a tag globally and not a tag Linked to a joke.

so for example you have a joke on the website and underneath the joke you see some tags associated with the joke. you have the option to press on a tag and Like it.
This will then tell the system you thought the Tag is a good Tag for the joke.

but with some small tweaks this could function perfectly when we finally implement the Tag feature.

great job everything works as expected.
i would rename the LikedTags.
and add some more info to the exception messages.

JunXingLiu committed:
Minor fixes

HellHunterMax reviewed (changes requested):
The model LikedTags should imo be renamed to LikeForTag.
Its not a LikedTags (the s making it multiple) anymore.
So there should be some more changes all over but just the name. and also controller names and stuff. @chrisK00  what do you think ?

JunXingLiu commented:
> The model LikedTags should imo be renamed to LikeForTag. Its not a LikedTags (the s making it multiple) anymore. So there should be some more changes all over but just the name. and also controller names and stuff. @chrisK00 what do you think ?

Should I change it for Repo and Services? from LikedTagsRepository -> LikedTagRepository

HellHunterMax commented:
> > The model LikedTags should imo be renamed to LikeForTag. Its not a LikedTags (the s making it multiple) anymore. So there should be some more changes all over but just the name. and also controller names and stuff. @chrisK00 what do you think ?
> 
> Should I change it for Repo and Services? from LikedTagsRepository -> LikedTagRepository

Repo, services, controller and model.
LikeForTagsRepository in my opinion... and change the model to a LikeForTag so then everything is named the same way.

JunXingLiu committed:
minor fixes

JunXingLiu commented:
> > > The model LikedTags should imo be renamed to LikeForTag. Its not a LikedTags (the s making it multiple) anymore. So there should be some more changes all over but just the name. and also controller names and stuff. @chrisK00 what do you think ?
> > 
> > 
> > Should I change it for Repo and Services? from LikedTagsRepository -> LikedTagRepository
> 
> Repo, services, controller and model. LikeForTagsRepository in my opinion... and change the model to a LikeForTag so then everything is named the same way.

I changed it. Hmm should the DB table be changed as well or that can remain?

JunXingLiu commented:
Crap, that's weird. It compiled fine on my computer... how do I fix those reference errors?

JunXingLiu committed:
extensions named fixes

HellHunterMax commented:
> > > > The model LikedTags should imo be renamed to LikeForTag. Its not a LikedTags (the s making it multiple) anymore. So there should be some more changes all over but just the name. and also controller names and stuff. @chrisK00 what do you think ?
> > > 
> > > 
> > > Should I change it for Repo and Services? from LikedTagsRepository -> LikedTagRepository
> > 
> > 
> > Repo, services, controller and model. LikeForTagsRepository in my opinion... and change the model to a LikeForTag so then everything is named the same way.
> 
> I changed it. Hmm should the DB table be changed as well or that can remain?

Yes the database table name should be the same as the model.

JunXingLiu committed:
fixed database table name and renamed LikedTag -> LikeForTag

JunXingLiu commented:
> > > > > The model LikedTags should imo be renamed to LikeForTag. Its not a LikedTags (the s making it multiple) anymore. So there should be some more changes all over but just the name. and also controller names and stuff. @chrisK00 what do you think ?
> > > > 
> > > > 
> > > > Should I change it for Repo and Services? from LikedTagsRepository -> LikedTagRepository
> > > 
> > > 
> > > Repo, services, controller and model. LikeForTagsRepository in my opinion... and change the model to a LikeForTag so then everything is named the same way.
> > 
> > 
> > I changed it. Hmm should the DB table be changed as well or that can remain?
> 
> Yes the database table name should be the same as the model.

Done!

HellHunterMax commented:
I just created the UserJokeTag that is actually needed for this issue.
Please review #207 Then we can edit this one to use UserJokeTag for liking.

JunXingLiu commented:
> I just created the UserJokeTag that is actually needed for this issue. Please review #207 Then we can edit this one to use UserJokeTag for liking.

Should I modify my code so it uses UserJokeTag? or that is another issue?

HellHunterMax commented:
Yes that would be best imo I'll merge to  dev.

JunXingLiu committed:
Merge branch 'dev' into feature/#90-user-liked-tags

JunXingLiu committed:
Implemented liking and unliking UserJokeTag

Combined Like and Unlike to a toggle.

JunXingLiu commented:
> Yes that would be best imo I'll merge to dev.

I think I got it working. Please take a look.

HellHunterMax commented on Server/JokesOnYou.Web.Api/Models/Request/LikeForTagCreateDto.cs:
```diff
@@ -0,0 +1,8 @@
+
+namespace JokesOnYou.Web.Api.Models.Request
+{
+    public class LikeForTagCreateDto
```
This is not being used so can be removed ?

HellHunterMax commented on Server/JokesOnYou.Web.Api/Models/Request/UnlikedTagCreateDto.cs:
```diff
@@ -0,0 +1,12 @@
+
+namespace JokesOnYou.Web.Api.Models.Request
+{
+    public class UnlikedTagCreateDto
```
same goes for this one. not being used ?

HellHunterMax commented on Server/JokesOnYou.Web.Api/Services/LikeForTagsService.cs:
```diff
@@ -0,0 +1,82 @@
+                }
+            }
+
+            var tagReplyDto = _mapper.Map<TagReplyDto>(tag);
```
This would return the likes from the tag. while what we need are the likes from the UserJokeTag.

JunXingLiu replied:
on line 77: tagReplyDto.Likes = userJokeTag.Likes
Do you think this will be okay?

HellHunterMax replied:
yes

HellHunterMax commented on Server/JokesOnYou.Web.Api/Services/TagService.cs:
```diff
@@ -60,15 +63,17 @@ public async Task<TagReplyDto> CreateTagAsync(TagCreateDto tagCreateDto, string
                     throw new AppException($"error: Failed saving {nameof(userJokeTag)} for the {nameof(tag)}: {tagCreateDto.Name}");
                 }
+
+                tagReplyDto = await _likeForTagsService.ToggleLikeForTag(userJokeTag, userId); // add initial like to LikeForTag
```
Lets put this duplicate line Under the if/else statement.

HellHunterMax commented on Server/JokesOnYou.Web.Api/Services/TagService.cs:
```diff
@@ -16,24 +16,27 @@ public class TagService : ITagService
                 tag = _mapper.Map<Tag>(tagCreateDto);
                 tag.OwnerId = userId;
-                tag.Likes = 1;
+                tag.Likes = 0;
```
we can remove this line as an int defaults to 0.

HellHunterMax commented on Server/JokesOnYou.Web.Api/Repositories/LikeForTagsRepository.cs:
```diff
@@ -0,0 +1,33 @@
+
+        public async Task AddLikedTag(LikeForTag likedTag) => await _context.LikeForTags.AddAsync(likedTag);
+
+        public async Task<IEnumerable<LikeForTag>> GetLikedTagByUserId(string userId) =>
```
Lets call this GetLikedTagsByUserId. The Added S telling us that this is multiple.

And TBH we did it wrong in all our Repos, Services and Controllers. it should just be called GetByUserId as we are inside the LikeForTags Repository we will of course return a LikeForTag would be weird to return something else.

HellHunterMax commented on Server/JokesOnYou.Web.Api/Controllers/LikeForTagsController.cs:
```diff
@@ -0,0 +1,44 @@
+                tagReplyDto = await _likeForTagsService.ToggleLikeForTag(userJokeTag, userId);
+            }
+
+            return tagReplyDto;
```
Maybe we should just return An Ok or a NoContent() depending on the like or Unlike ?

JunXingLiu committed:
changed controller input to jokeid and tagid

HellHunterMax reviewed (approved):
Great job! Thanks for all the edits!

HellHunterMax commented:
I see I still had some unpushed comments ready. You could fix those as well if you want to make it complete.

JunXingLiu commented:
> I see I still had some unpushed comments ready. You could fix those as well if you want to make it complete.

I'll get those out of the way
